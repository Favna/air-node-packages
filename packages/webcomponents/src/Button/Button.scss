@import '../styles/vars';

.button {
  @extend %fontNormal;
  background-color: $blue;
  border-radius: 0;
  box-shadow: none;
  float: right;
  height: 3.2rem;
  line-height: initial;
  margin-right: 1%;

  &Label {
    @extend %fontNormal;
    color: $white;
    font-size: 1rem;
    text-transform: none;
  }

  &Primary {
    border: 2px solid $blue;
  }

  &Secondary {
    background-color: $yellow;
    border: 2px solid $yellow;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }

  &Outlined {
    background-color: $white;
    border: 2px solid $blue;
    color: $blue;

    > span {
      color: $blue;
    }

    &:hover {
      border-width: 2px;
    }

    &Secondary {
      background-color: $white;
      border: 2px solid $darkOrange;
      color: $textBlack;

      > span {
        color: $textBlack;
      }
    }
  }

  &Disabled {
    background-color: $lightGrey;
    border: transparent;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }
}
