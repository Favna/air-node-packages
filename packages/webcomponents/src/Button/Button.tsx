/**
 * @class Button
 */

import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import css from './Button.scss';
import classnames from 'classnames';
import React, { FC, ReactNode, memo } from 'react';
import { customCss, dataQa } from '../constants';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
  /** Custom CSS classes to pass to the button */
  customclasses?: customCss;
  /** The action that should be triggered when clicking the button */
  onClick(): any;
}

const useStyles = makeStyles((theme: Theme) => createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } }),
);

/**
 * Creates a button using pre-defined Rijkswatestaat styling
 *
 * ```jsx
 * import { Button } from '@rws-air/webcomponents'
 *
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */
export const Button: FC<MUIButtonProps<'button', ButtonProps>> = props => {
  const classes = useStyles();

  return (
    <MUIButton
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: css.buttonLabel,
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow),
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
};

export default memo(Button);