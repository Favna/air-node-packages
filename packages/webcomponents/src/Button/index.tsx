import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import clsx from 'clsx';
import React, { forwardRef, memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @remark Ref is forwarded to the Material-UI button to allow for wrapping this button with a Tooltip
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */

export const Button = memo(
  forwardRef<HTMLButtonElement, MUIButtonProps<'button', ButtonProps>>((props, ref) => {
    const classes = useStyles();

    return (
      <MUIButton
        {...props}
        ref={ref}
        data-qa={props['data-qa']}
        onClick={props.onClick}
        variant={props.variant}
        color={props.color}
        disabled={props.disabled}
        className={clsx(props.customclasses)}
        classes={{
          root: css.button,
          label: clsx(css.buttonLabel, props.customlabelclasses),
          disabled: css.buttonDisabled,
          containedPrimary: clsx(css.buttonPrimary, classes.buttonShadow),
          containedSecondary: clsx(css.buttonSecondary, classes.buttonShadow),
          outlined: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedPrimary: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedSecondary: clsx(css.buttonOutlinedSecondary, classes.buttonShadow)
        }}
        href={undefined}
      >
        {props.label}
      </MUIButton>
    );
  })
);
