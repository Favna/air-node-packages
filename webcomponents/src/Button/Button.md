## Usage of the Button component:

**Example:**

```tsx
import React, { FC } from 'react'

import { Button } from '@rws-air/webcomponents'

const Example: FC = () => (
  <Button variant='contained' color='primary' label='button' onClick={() => console.log('void')} />
)
```

**Primary Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button variant='contained' color='primary' label='I am primary!' onClick={() => console.log('you clicked the basic button!')} />
```

**Disabled Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button variant='contained' color='primary' label='I am disabled!' disabled />
```

**Secondary Color button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button
  variant='contained' color='secondary' label='I am secondary!'
  onClick={() => console.log('you clicked the secondary button!')} 
/>
```

**Outline Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button
  variant='outlined' color='primary' label='I am outlined!'
  onClick={() => console.log('you clicked the outlined button!')} 
/>
```