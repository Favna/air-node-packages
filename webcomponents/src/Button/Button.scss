@import '../styles/vars.scss';

.rwsButton {
  @include font-normal;
  background-color: $blue !important;
  border-radius: 0        !important;
  box-shadow: none        !important;
  float: right            !important;
  height: 3.2rem          !important;
  line-height: initial    !important;
  margin-right: 1%        !important;
}

.rwsButtonLabel {
  @include font-normal;
  color: $white           !important;
  font-size: 1rem         !important;
  font-weight: bold       !important;
  text-transform: none    !important;
}

button[class*="Secondary"] {
  background-color: $yellow !important;
}

button:disabled {
  background-color: $light-grey !important;
}

button[class*="outlined"] {
  background-color: $white !important;
  border: 2px solid $blue !important;
  color: $text-black !important;

  > span {
    color: $text-black !important;
  }
}