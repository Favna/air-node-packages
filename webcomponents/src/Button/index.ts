export { default as Button } from './Button';
export { RWSButtonProps as ButtonProps } from './Button';