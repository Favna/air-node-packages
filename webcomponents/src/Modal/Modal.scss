@import '../styles/vars';

%modal-button-extendible {
  width: 11.5rem;
}

.modal {
  > div > div {
    max-width: unset;
    min-width: 45vw;
  }

  &Actions {
    margin-bottom: 3vh;
    margin-right: 5vw;
    padding-right: 1.5rem;
  }

  &Body {
    margin: 0 5vw;
    overflow: hidden;

    &Small {
      max-height: 10rem;
    }
  }

  &Button {
    &Cancel {
      @extend %modal-button-extendible;
      background-color: $white;
      border-width: 2px;

      > span {
        color: $blue;
      }
    }

    &Ok {
      @extend %modal-button-extendible;
      margin-left: 1%;
    }
  }
}

.title {
  margin-left: 5vw;
  margin-right: 5vw;
  padding-top: 3vw;

  > * {
    color: $dark-blue;
    font-size: 1.1rem;
    font-weight: bolder;
  }

  &Icon {
    box-sizing: content-box;
    padding-bottom: 0.6rem;
    padding-right: 0.8rem;

    &CheckCircle {
      color: $green;
    }

    &Offset {
      margin-left: calc(5vw - 2.5rem);

      &Header > h6 {
        align-items: center;
        display: flex;
      }
    }
  }
}
