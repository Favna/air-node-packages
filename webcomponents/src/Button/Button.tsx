/**
 * @class Button
 */

import React, { FC, ReactNode } from 'react'
import classnames from 'classnames';
import css from './Button.scss';
import { Button as MUIButton } from '@material-ui/core';
import { ButtonProps } from '@material-ui/core/Button';

import './Button.scss';

export type RWSButtonProps = {
    /** The variant of the button */
    variant: 'text' | 'outlined' | 'contained' | undefined;
    /** The color type of the button */
    color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
    /** The label to be displayed inside the button */
    label: ReactNode;
    /** Whether this button should be disabled */
    disabled?: boolean;
    /** Any extra CSS classes to be passed to the button */
    customClasses?: string | string[];

    /** The action that should be triggered when clicking the button */
    onClick(): any;
};

export const Button: FC<ButtonProps<'button', RWSButtonProps>> = props => (
    <MUIButton
        onClick={props.onClick}
        variant={props.variant}
        color={props.color}
        disabled={props.disabled}
        className={classnames(props.customClasses)}
        classes={{ root: css.rwsButton, label: css.rwsButtonLabel }}
        href={undefined}
    >
        {props.label}
    </MUIButton>
)

export default Button;