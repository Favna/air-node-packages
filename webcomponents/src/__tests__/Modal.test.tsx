import CheckCircle from '@material-ui/icons/CheckCircle';
import WarningIcon from '@material-ui/icons/Warning';
import Button from '../Button/Button';
import { shallow, ShallowWrapper } from 'enzyme';
import Modal, { ModalProps } from '../Modal/Modal';
import React, { Fragment, ReactNode } from 'react';
import { OptionalKeys } from 'utility-types';

let modal: ShallowWrapper<ModalProps, any>;
const closeAction = jest.fn();
const confirmAction = jest.fn();

const setup = (
  isOpen: boolean = false,
  isSimpleModal: boolean = false,
  props?: Pick<ModalProps, OptionalKeys<ModalProps>>,
  dialogContent: ReactNode = <Fragment />
) => (
    modal = shallow<ModalProps>(<Modal
      {...props}
      topic='Awesome Topic'
      dialogContent={dialogContent}
      closeAction={closeAction}
      confirmAction={confirmAction}
      open={isOpen}
      isSimpleModal={isSimpleModal}
      modalqas={{
        content: 'modal-content',
        modal: 'modal',
        title: 'modal-title',
        actionCancel: 'modal-cancel-button',
        actionConfirm: 'modal-confirm-button',
        actions: 'modal-actions',
      }}
    />)
  );

describe('Render Testing', () => {
  test('should include mandatory props', () => {
    setup();
    expect(modal.find(Button).length).toBe(0);
    expect(modal.prop('disableBackdropClick')).toBeTruthy();
    expect(modal.prop('disableEscapeKeyDown')).toBeTruthy();
  });

  test('should pass all data-qa\'s', () => {
    setup(true, true);

    expect(modal.find('[data-qa="modal"]').exists()).toBeTruthy();
    expect(modal.find('[data-qa="modal-title"]').exists()).toBeTruthy();
    expect(modal.find('[data-qa="modal-cancel-button"]').exists()).toBeTruthy();
    expect(modal.find('[data-qa="modal-confirm-button"]').exists()).toBeTruthy();
    expect(modal.find('[data-qa="modal-actions"]').exists()).toBeTruthy();
  });

  test('should include optional props', () => {
    setup(
      true, true,
      {
        okButtonText: 'Awesome Ok Text',
        cancelButtonText: 'Awesome Cancel Text',
        modalType: 'warning',
      }
    );

    expect(modal.find('[data-qa="modal-title"]').childAt(1).text()).toBe('Awesome Topic');
    expect(modal.find(Button).at(0).prop('label')).toBe('Awesome Cancel Text');
    expect(modal.find(Button).at(1).prop('label')).toBe('Awesome Ok Text');
  });

  test('should show CheckCircle when modalType == check', () => {
    setup(true, true);

    expect(modal.exists()).toBeTruthy();
    expect(modal.find(CheckCircle).exists()).toBeTruthy();
  });

  test('should show Warning Triangle when modalType == check', () => {
    setup(true, true, { modalType: 'warning' });

    expect(modal.find(WarningIcon).exists()).toBeTruthy();
  });
});

describe('Action Testing', () => {
  beforeAll(() => setup(true, true));
  test('Ok button should work', () => {
    const okButton = modal.find('[data-qa="modal-confirm-button"]');
    okButton.simulate('click');
    expect(confirmAction).toHaveBeenCalled();
    expect(confirmAction).toHaveBeenCalledTimes(1);
  });

  test('Cancel button should work', () => {
    const cancelButton = modal.find('[data-qa="modal-cancel-button"]');
    cancelButton.simulate('click');
    expect(closeAction).toHaveBeenCalled();
    expect(closeAction).toHaveBeenCalledTimes(1);
  });
});

describe('Content Testing', () => {
  const content = <div>Content</div>;
  beforeAll(() => setup(true, false, {}, content));

  test('should ', () => {
    expect(modal.find('[data-qa="modal-content"]').children().exists()).toBeTruthy();
    expect(modal.find('[data-qa="modal-content"]').children()).toStrictEqual(shallow(content));
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    setup();
    expect(modal).toMatchSnapshot();
  });

  test('Optional Props', () => {
    setup(
      true, true,
      {
        okButtonText: 'Awesome Ok Text',
        cancelButtonText: 'Awesome Cancel Text',
        modalType: 'warning',
      }
    );

    expect(modal).toMatchSnapshot();
  });

  test('Open and simple modal', () => {
    setup(true, true);
    expect(modal).toMatchSnapshot();
  });
});