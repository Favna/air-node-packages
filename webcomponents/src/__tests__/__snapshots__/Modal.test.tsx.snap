// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Open and simple modal 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "root": undefined,
    }
  }
  data-qa="modal"
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onKeyPress={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    className="undefined"
    data-qa="modal-title"
  >
    <CheckCircleIcon
      className=""
      color="inherit"
    />
    Awesome Topic
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="undefined"
    classes={
      Object {
        "root": undefined,
      }
    }
    data-qa="modal-content"
  />
  <WithStyles(ForwardRef(DialogActions))
    classes={
      Object {
        "root": undefined,
      }
    }
    data-qa="modal-actions"
  >
    <Button
      color="primary"
      data-qa="modal-cancel-button"
      label="cancel"
      onClick={[Function]}
      variant="outlined"
    />
    <Button
      color="primary"
      data-qa="modal-confirm-button"
      label="ok"
      onClick={[Function]}
      variant="contained"
    />
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "root": undefined,
    }
  }
  data-qa="modal"
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onKeyPress={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    className="undefined"
    data-qa="modal-title"
  >
    <WarningIcon
      className=""
      color="error"
    />
    Awesome Topic
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="undefined"
    classes={
      Object {
        "root": undefined,
      }
    }
    data-qa="modal-content"
  />
  <WithStyles(ForwardRef(DialogActions))
    classes={
      Object {
        "root": undefined,
      }
    }
    data-qa="modal-actions"
  >
    <Button
      color="primary"
      data-qa="modal-cancel-button"
      label="Awesome Cancel Text"
      onClick={[Function]}
      variant="outlined"
    />
    <Button
      color="primary"
      data-qa="modal-confirm-button"
      label="Awesome Ok Text"
      onClick={[Function]}
      variant="contained"
    />
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Dialog))
  classes={
    Object {
      "root": undefined,
    }
  }
  data-qa="modal"
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onKeyPress={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    className=""
    data-qa="modal-title"
  >
    Awesome Topic
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className=""
    classes={
      Object {
        "root": undefined,
      }
    }
    data-qa="modal-content"
  />
</WithStyles(ForwardRef(Dialog))>
`;
